yourbench/utils/inference_engine.py

```diff
@@ -1,122 +1,301 @@
+    )
+    return ""
+
+
+async def _run_inference_async_helper(
+    models: List[Model], inference_calls: List[InferenceCall]
+) -> Dict[str, List[str]]:
+    """
+    Launch tasks for each (model, inference_call) pair in parallel, respecting concurrency.
+
+    Returns:
+        Dict[str, List[str]]: A dictionary keyed by model.model_name, each value
+        is a list of responses (strings) in the same order as 'inference_calls'.
+    """
+    logger.info("Starting asynchronous inference with per-model concurrency control.")
+
+    # Instead of a single global concurrency, create a semaphore per model based on model.max_concurrent_requests
+    model_semaphores: Dict[str, asyncio.Semaphore] = {}
+    for model in models:
+        # If not specified, default to something reasonable like 1
+        concurrency = max(model.max_concurrent_requests, 1)
+        semaphore = asyncio.Semaphore(concurrency)
+        model_semaphores[model.model_name] = semaphore
+        logger.debug("Created semaphore for model='{}' with concurrency={}", model.model_name, concurrency)
+
+    tasks = []
+    # We'll build tasks in an order that ensures each model gets a contiguous
+    # slice in the final results.
+    for model in models:
+        semaphore = model_semaphores[model.model_name]
+        for call in inference_calls:
+            tasks.append(_retry_with_backoff(model, call, semaphore))
+
+    logger.info("Total tasks scheduled: {}  (models={}  x  calls={})", len(tasks), len(models), len(inference_calls))
+
+    # Run them all concurrently
+    results = await tqdm_asyncio.gather(*tasks)
+    logger.success("Completed parallel inference for all models.")
+
+    # Re-map results back to {model_name: [list_of_responses]}
+    responses: Dict[str, List[str]] = {}
+    idx = 0
+    n_calls = len(inference_calls)
+    for model in models:
+        slice_end = idx + n_calls
+        model_responses = results[idx:slice_end]
+        responses[model.model_name] = model_responses
+        idx = slice_end
+
+    # Optional debug: confirm each model's result count
+    for model in models:
+        logger.debug("Model '{}' produced {} responses.", model.model_name, len(responses[model.model_name]))
+
+    return responses
+
+
+def _load_models(base_config: Dict[str, Any], step_name: str) -> List[Model]:
+    """
+    Load only the models assigned to this step from the config's 'model_list' and 'model_roles'.
+    If no model role is defined for the step, use the first model from model_list.
+    """
+    all_configured_models = base_config.get("model_list", [])
+    role_models = base_config.get("model_roles", {}).get(step_name, [])
+
+    # If no role models are defined for this step, use the first model from model_list
+    if not role_models and all_configured_models:
+        logger.info(
+            "No models defined in model_roles for step '{}'. Using the first model from model_list: {}",
```

Comment:
We could want to enforce the user explicitely setting params by removing defaults

Reply:
wanted to make it as simple as possible for a user to run the pipeline, so if they were to just add a single model, it would be easy for us to use that one model to perform inference. lmkyt, I think this is an important default